rdrf/rdrf/forms/widgets/settings_widgets.py

```diff
@@ -1,242 +1,151 @@
-            self.generate_input('left_label', _('Left label'), parsed),
-            self.generate_input('right_label', _('Right label'), parsed),
-            self.generate_input('step', _('Step'), parsed),
-        ]
-        return "<br/>".join(rows)
+    def get_allowed_fields(self):
+        pass
 
-    def render(self, name, value, attrs=None, renderer=None):
-        parsed = {}
+    def parse_value(self, value):
+        self.parsed = {}
         try:
-            parsed = json.loads(value)
+            self.parsed = json.loads(value)
+            self.parsed = {k: v for k, v in self.parsed.items() if k in self.get_allowed_fields()}
         except Exception:
             pass
 
-        html = """
-             <div style="display:inline-grid" id="id_{name}">
-                {inputs}
-                <input type="hidden" name="{name}" value='{value}'/>
-             </div>""".format(inputs=self.generate_inputs(parsed), name=name, value=value)
-        javascript = """
-            function saveJSON() {
-                var inputs = $('#id_%s input[type!=hidden]');
-                var obj = {};
-                for (var i = 0; i < inputs.length; i++) {
-                    if (inputs[i].value !='' && inputs[i].value.trim() != '') {
-                        obj[inputs[i].name] = inputs[i].value;
-                    }
-                }
-                $("input[name='%s']").val(JSON.stringify(obj));
-            }
-            saveJSON();
-        """ % (name, name)
-        return mark_safe(f"""
-            {html}
-            <script>
-                {javascript}
-            </script>""")
+    def _update_input(self, name, title, info=None, onchange=None):
+        value = self.parsed.get(name, '')
+        self.parsed[name] = {
+            'name': name,
+            'value': value,
+            'title': title,
+            'info': info,
+            'onchange': onchange or 'saveJSON()'
+        }
+
+    def generate_text_input(self, name, title, **kwargs):
+        self._update_input(name, title, kwargs.get('info'), kwargs.get('onchange'))
+        self.parsed[name]['input_type'] = 'text'
+
+    def generate_checkbox_input(self, name, title, **kwargs):
+        self._update_input(name, title, kwargs.get('info'), kwargs.get('onchange'))
+        self.parsed[name]['input_type'] = 'checkbox'
+        self.parsed[name]['checked'] = kwargs.get('checked')
+
+    def generate_select_input(self, name, title, **kwargs):
+        self._update_input(name, title, kwargs.get('info'), kwargs.get('onchange'))
+        self.parsed[name]['input_type'] = 'select'
+        self.parsed[name]['options'] = kwargs.get('options', [])
+
+    def generate_inputs(self):
+        pass
+
+    def render(self, name, value, attrs=None, renderer=None):
+        self.parse_value(value)
+        self.generate_inputs()
+        return self.render_widget(name, value, renderer)
+
+    def render_widget(self, name, value, renderer=None, javascript=None):
```

Comment:
Can we make `javascript` an instance variable instead? We default it to `None` then classes that have extra js will just define the js on the class instead of having to override `render` just to pass the javascript in (as it is done currently in `DurationWidgetSettings`)?

Reply:
Updated. Still need to override render though as the widget name is a parameter to the render method.